### ISSUE TITLE
ctypes - loading external DLL (not bundled) fails (not able to load DLL from current working directory)

### ISSUE THREAD

**Issue by HeartUnchange** [pyinstaller/pyinstaller#1609]
i have a problem again,I build py file which use ctypes and pyqt4 modules.but it failed,the error information is here:

    D:\fuck\bp>pyinstaller --onefile bp.py
    55 INFO: PyInstaller: 3.0
    55 INFO: Python: 2.7.10
    57 INFO: Platform: Windows-7-6.1.7601-SP1
    58 INFO: wrote D:\fuck\bp\bp.spec
    60 INFO: UPX is not available.
    61 INFO: Extending PYTHONPATH with D:\fuck\bp
    61 INFO: checking Analysis
    193 INFO: Building because D:\fuck\bp\bp.py changed
    194 INFO: Initializing module dependency graph...
    198 INFO: Initializing module graph hooks...
    253 INFO: running Analysis out00-Analysis.toc
    257 INFO: Adding Microsoft.VC90.CRT to dependent assemblies of final executable
      required by e:\python27\python.exe
    354 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_1f
    c8b3b9a1e18e3b_9.0.30729.1_none_8550c6b5d18a9128.manifest
    355 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_1f
    c8b3b9a1e18e3b_9.0.30729.4148_none_f47e1bd6f6571810.manifest
    356 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_1f
    c8b3b9a1e18e3b_9.0.30729.4940_none_f47ed0f6f6564d90.manifest
    357 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_1f
    c8b3b9a1e18e3b_9.0.30729.6161_none_f480bfaef65491a5.manifest
    460 INFO: Searching for assembly x86_Microsoft.VC90.CRT_1fc8b3b9a1e18e3b_9.0.307
    29.6161_none ...
    460 INFO: Found manifest C:\Windows\WinSxS\Manifests\x86_microsoft.vc90.crt_1fc8
    b3b9a1e18e3b_9.0.30729.6161_none_50934f2ebcb7eb57.manifest
    462 INFO: Searching for file msvcr90.dll
    462 INFO: Found file C:\Windows\WinSxS\x86_microsoft.vc90.crt_1fc8b3b9a1e18e3b_9
    .0.30729.6161_none_50934f2ebcb7eb57\msvcr90.dll
    463 INFO: Searching for file msvcp90.dll
    463 INFO: Found file C:\Windows\WinSxS\x86_microsoft.vc90.crt_1fc8b3b9a1e18e3b_9
    .0.30729.6161_none_50934f2ebcb7eb57\msvcp90.dll
    463 INFO: Searching for file msvcm90.dll
    465 INFO: Found file C:\Windows\WinSxS\x86_microsoft.vc90.crt_1fc8b3b9a1e18e3b_9
    .0.30729.6161_none_50934f2ebcb7eb57\msvcm90.dll
    568 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_1f
    c8b3b9a1e18e3b_9.0.30729.1_none_8550c6b5d18a9128.manifest
    570 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_1f
    c8b3b9a1e18e3b_9.0.30729.4148_none_f47e1bd6f6571810.manifest
    572 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_1f
    c8b3b9a1e18e3b_9.0.30729.4940_none_f47ed0f6f6564d90.manifest
    573 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_1f
    c8b3b9a1e18e3b_9.0.30729.6161_none_f480bfaef65491a5.manifest
    574 INFO: Adding redirect Microsoft.VC90.CRT version (9, 0, 21022, 8) -> (9, 0,
    30729, 6161)
    764 INFO: Analyzing D:\fuck\bp\bp.py
    5022 INFO: Processing pre-find module path hook   distutils
    9170 INFO: Processing pre-find module path hook   site
    9185 INFO: site: retargeting to fake-dir 'e:\\python27\\lib\\site-packages\\PyIn
    staller\\fake-modules'
    12445 INFO: Looking for import hooks ...
    12451 INFO: Processing hook   hook-distutils.py
    12479 INFO: Processing hook   hook-sysconfig.py
    12480 INFO: Processing hook   hook-xml.py
    12667 INFO: Processing hook   hook-win32com.client.py
    12667 INFO: Processing hook   hook-httplib.py
    12669 INFO: Processing hook   hook-pydoc.py
    12671 INFO: Excluding import 'Tkinter'
    12671 INFO: Processing hook   hook-PyQt4.py
    12674 WARNING: Hidden import 'PyQt4._qt' not found (probably old hook)
    12674 INFO: Processing hook   hook-encodings.py
    13208 INFO: Processing hook   hook-_tkinter.py
    13676 INFO: checking Tree
    13916 INFO: checking Tree
    14021 INFO: Processing hook   hook-PyQt4.QtGui.py
    15458 WARNING: Hidden import 'PyQt4._qt' not found (probably old hook)
    15459 INFO: Processing hook   hook-pywintypes.py
    15754 INFO: Processing hook   hook-setuptools.py
    15756 INFO: Processing hook   hook-xml.sax.py
    15774 INFO: Processing hook   hook-PyQt4.QtCore.py
    15865 WARNING: Hidden import 'PyQT4._qt' not found (probably old hook)
    15866 INFO: Processing hook   hook-win32com.py
    16072 INFO: Processing hook   hook-pythoncom.py
    16413 INFO: Looking for ctypes DLLs
    Traceback (most recent call last):
      File "E:\python27\Scripts\pyinstaller-script.py", line 9, in <module>
        load_entry_point('PyInstaller==3.0', 'console_scripts', 'pyinstaller')()
      File "e:\python27\lib\site-packages\PyInstaller\__main__.py", line 99, in run
        run_build(opts, spec_file, pyi_config)
      File "e:\python27\lib\site-packages\PyInstaller\__main__.py", line 47, in run_
    build
        PyInstaller.building.build_main.main(pyi_config, spec_file, **opts.__dict__)
    
      File "e:\python27\lib\site-packages\PyInstaller\building\build_main.py", line
    737, in main
        build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'
    ))
      File "e:\python27\lib\site-packages\PyInstaller\building\build_main.py", line
    680, in build
        exec(text, spec_namespace)
      File "<string>", line 16, in <module>
      File "e:\python27\lib\site-packages\PyInstaller\building\build_main.py", line
    207, in __init__
        self.__postinit__()
      File "e:\python27\lib\site-packages\PyInstaller\building\datastruct.py", line
    156, in __postinit__
        self.assemble()
      File "e:\python27\lib\site-packages\PyInstaller\building\build_main.py", line
    468, in assemble
        ctypes_binaries, ctypes_warnings = scan_code_for_ctypes(co)
      File "e:\python27\lib\site-packages\PyInstaller\depend\utils.py", line 164, in
     scan_code_for_ctypes
        nested_binaries, nested_warnings = scan_code_for_ctypes(c)
      File "e:\python27\lib\site-packages\PyInstaller\depend\utils.py", line 181, in
     scan_code_for_ctypes
        elif binary != os.path.basename(binary):
      File "e:\python27\lib\ntpath.py", line 208, in basename
        return split(p)[1]
      File "e:\python27\lib\ntpath.py", line 180, in split
        d, p = splitdrive(p)
      File "e:\python27\lib\ntpath.py", line 116, in splitdrive
        normp = p.replace(altsep, sep)
    AttributeError: 'tuple' object has no attribute 'replace'

the py,spec and toc file have been uploade to here,
https://github.com/HeartUnchange/for_issue/tree/master/pyinstaller/bp

**Comment by HeartUnchange** [comment]
i print the binary in function scan_code_for_ctypes that is contained in File "e:\python27\lib\site-packages\PyInstaller\depend\utils.py".

and get these:
    12503 INFO: Looking for ctypes DLLs
    []
    []
    ['kernel32.dll']
    []
    []
    ['kernel32.dll', 'kernel32.dll', 'kernel32.dll']
    ['kernel32.dll']
    []
    []
    ['user32.dll']
    ['D:/fuck/bp/library.dll']
    []
    []
    []
    []
    [('D:/fuck/bp/library.dll', 'D:/fuck/bp/library.dll', 'BINARY')]
    Traceback (most recent call last):

**Comment by HeartUnchange** [comment]
it didn't happen again,amazing,lol.

**Comment by HeartUnchange** [comment]
i found if load dll happen in a class initlazation,pyinstaller cann't build.
like this:
    class Canvas(QtGui.QWidget):
        def __init__(self,parent=None,size_w=0,size_h=0):
            super(Canvas, self).__init__(parent)
            self.lib = ctypes.CDLL('D:/fuck/bp/library.dll')

**Comment by matysek** [comment]
> self.lib = ctypes.CDLL('D:/fuck/bp/library.dll')

Does it work when you use backslashes in the path?

**Comment by htgoebel** [comment]
> self.lib = ctypes.CDLL('D:/fuck/bp/library.dll')

PyInstaller can not handle absolute pates here. The absolute path will be there in the frozen app, too, but the dll will not.

@matysek: I' thinking about checking this in scan_code_for_ctypes and *fail* is an absolute path occurs as the first. element.

**Comment by ryansturmer** [comment]
I have found also that ctypes dlls loaded from within a class cause this crash.

**Comment by htgoebel** [comment]
So please provide a **minimal** example showing this issue. The stuff at https://github.com/HeartUnchange/for_issue/tree/master/pyinstaller/bp is much to long to spot the problem area.

**Comment by matysek** [comment]
@HeartUnchange I see similar issue when trying to use vlc python bindings: https://wiki.videolan.org/PythonBinding/
* source
```
Usage: openlp\core\ui\media\vendor\vlc.py <movie_filename>
Once launched, type ? for help.

Build date: Wed Jun 25 13:46:01 2014 (0x0)
LibVLC version: 2.2.1 Terry Pratchett (Weatherwax) (0x2020100)
LibVLC compiler: gcc version 4.9.1 (GCC)
Plugin path: C:\Program Files (x86)\VideoLAN\VLC
```
* frozen
```
LOADER: extracted pyimod01_os_path
LOADER: callfunction returned...
LOADER: extracted pyimod02_archive
LOADER: callfunction returned...
LOADER: extracted pyimod03_importers
LOADER: callfunction returned...
LOADER: Installing PYZ archive with Python modules.
LOADER: PYZ archive: out00-PYZ.pyz
LOADER: Running pyiboot01_bootstrap.py
LOADER: Running vlc.py
Traceback (most recent call last):
  File "<string>", line 172, in <module>
  File "<string>", line 145, in find_lib
  File "<string>", line 124, in __init__
  File "C:\tools\python3-x86_32\Lib\ctypes\__init__.py", line 351, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: [WinError 126] Module not found
vlc returned -1
LOADER: OK.
LOADER: Cleaning up Python interpreter.
```

**Comment by springermac** [comment]
I think I may have found the problem.

In the bootloader we call SetDllDirectory with our extraction path. For some reason this seems to prevent loading a dll from the current directory.
If SetDllDirectory is called with a Null argument to reset the dll search path loading a dll from the current directory works.

Example based on vlc and the libvlc.dll.

Doesn't work:

    import ctypes
    import os

    os.chdir('C:\\Program Files\\VideoLan\\VLC\\')
    print(ctypes.CDLL('libvlc.dll'))

Works:

    import ctypes
    import os

    ctypes.windll.kernel32.SetDllDirectoryW(None)

    os.chdir('C:\\Program Files\\VideoLan\\VLC\\')
    print(ctypes.CDLL('libvlc.dll'))

**Comment by codewarrior0** [comment]
Right, that's a side-effect of SetDllDirectory. In the DLL search order, it replaces <cwd> with the given path. Since I added that call, someone has added a runtime hook that overrides `ctypes` to add the extraction dir to the search path, so the SetDllDirectory call can probably be removed now. (The other bootloader code that sets LD_LIBRARY_PATH could probably be removed too in light of the `ctypes` runtime hook.)

**Comment by htgoebel** [comment]
> so the SetDllDirectory call can probably be removed now.

I doubt this. The bootloader will need to load a lot of DLLs until it comes to execute the loader scripts. At least it requires to load libpython.

Further: Loading DLLs from the current directory is a security hole! Remember the Windows did this for along time, allowing for lots of attacks.

**Comment by matysek** [comment]
@htgoebel @codewarrior0 should this issue be then converted into a recipe?

**Comment by springermac** [comment]
Commit 6c1a02c7a3aafd925c3831c3e45688aeeabe6159 is the commit which broke loading DLLs from the current directory.

**Comment by matysek** [comment]
@HeartUnchange I'm sorry for misleading in this issue. Issue for your original bug report is in #1620.

**Comment by matysek** [comment]
> self.lib = ctypes.CDLL('D:/fuck/bp/library.dll')
> PyInstaller can not handle absolute pates here. The absolute path will be there in the frozen app, too, but the dll will not.

@htgoebel I created #1620 for this.

**Comment by HeartUnchange** [comment]
@matysek it's ok.I am very pleased that we can find so much through this issue,

**Comment by HeartUnchange** [comment]
@htgoebel  there is a minimal example i can provide:
[#1620](https://github.com/pyinstaller/pyinstaller/issues/1620)

**Comment by HeartUnchange** [comment]
@htgoebel 
there is a example on windows,

    #!/usr/bin/env python
    #-*- coding:utf-8 -*-

    import ctypes

    class test_class:
        def __init__(self):
            self.lib = ctypes.CDLL("library.dll")
        def life(self):
            self.lib.love()

    if __name__ == '__main__':
        test = test_class()
        test.life()

the source of library.dll is : 
  
    extern "C"{
        #include <stdio.h>
        int love(){
           printf("I still love the life" );
          return 0;
       }
    }

compile command :

    g++ -c -fPIC test.c -o test.o
    g++ -shared -Wl,-soname,library.dll -o library.dll test.o

the error information is :

    pyinstaller --onefile mini_bp.py
    1500 INFO: PyInstaller: 3.0
    1506 INFO: Python: 2.7.10
    1506 INFO: Platform: Windows-7-6.1.7601-SP1
    1509 INFO: wrote D:\fuck\bp\mini_bp.spec
    1513 INFO: UPX is not available.
    1539 INFO: Extending PYTHONPATH with D:\fuck\bp
    1539 INFO: checking Analysis
    1539 INFO: Building Analysis because out00-Analysis.toc is non existent
    1539 INFO: Initializing module dependency graph...
    1600 INFO: Initializing module graph hooks...
    1888 INFO: running Analysis out00-Analysis.toc
    2029 INFO: Adding Microsoft.VC90.CRT to dependent assemblies of final executable
    
      required by e:\python27\python.exe
    35225 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_
    1fc8b3b9a1e18e3b_9.0.30729.1_none_8550c6b5d18a9128.manifest
    35242 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_
    1fc8b3b9a1e18e3b_9.0.30729.4148_none_f47e1bd6f6571810.manifest
    35263 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_
    1fc8b3b9a1e18e3b_9.0.30729.4940_none_f47ed0f6f6564d90.manifest
    35279 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_
    1fc8b3b9a1e18e3b_9.0.30729.6161_none_f480bfaef65491a5.manifest
    35398 INFO: Searching for assembly x86_Microsoft.VC90.CRT_1fc8b3b9a1e18e3b_9.0.3
    0729.6161_none ...
    35398 INFO: Found manifest C:\Windows\WinSxS\Manifests\x86_microsoft.vc90.crt_1f
    c8b3b9a1e18e3b_9.0.30729.6161_none_50934f2ebcb7eb57.manifest
    35406 INFO: Searching for file msvcr90.dll
    35407 INFO: Found file C:\Windows\WinSxS\x86_microsoft.vc90.crt_1fc8b3b9a1e18e3b
    _9.0.30729.6161_none_50934f2ebcb7eb57\msvcr90.dll
    35409 INFO: Searching for file msvcp90.dll
    35410 INFO: Found file C:\Windows\WinSxS\x86_microsoft.vc90.crt_1fc8b3b9a1e18e3b
    _9.0.30729.6161_none_50934f2ebcb7eb57\msvcp90.dll
    35411 INFO: Searching for file msvcm90.dll
    35414 INFO: Found file C:\Windows\WinSxS\x86_microsoft.vc90.crt_1fc8b3b9a1e18e3b
    _9.0.30729.6161_none_50934f2ebcb7eb57\msvcm90.dll
    35539 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_
    1fc8b3b9a1e18e3b_9.0.30729.1_none_8550c6b5d18a9128.manifest
    35541 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_
    1fc8b3b9a1e18e3b_9.0.30729.4148_none_f47e1bd6f6571810.manifest
    35544 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_
    1fc8b3b9a1e18e3b_9.0.30729.4940_none_f47ed0f6f6564d90.manifest
    35545 INFO: Found C:\Windows\WinSxS\Manifests\x86_policy.9.0.microsoft.vc90.crt_
    1fc8b3b9a1e18e3b_9.0.30729.6161_none_f480bfaef65491a5.manifest
    35546 INFO: Adding redirect Microsoft.VC90.CRT version (9, 0, 21022, 8) -> (9, 0
    , 30729, 6161)
    36271 INFO: Analyzing D:\fuck\bp\mini_bp.py
    41858 INFO: Looking for import hooks ...
    41898 INFO: Processing hook   hook-httplib.py
    41909 INFO: Processing hook   hook-encodings.py
    43457 INFO: Looking for ctypes DLLs
    43488 WARNING: library python%s%s required via ctypes not found
    Traceback (most recent call last):
      File "E:\python27\Scripts\pyinstaller-script.py", line 9, in <module>
        load_entry_point('PyInstaller==3.0', 'console_scripts', 'pyinstaller')()
      File "e:\python27\lib\site-packages\PyInstaller\__main__.py", line 99, in run
        run_build(opts, spec_file, pyi_config)
      File "e:\python27\lib\site-packages\PyInstaller\__main__.py", line 47, in run_
    build
        PyInstaller.building.build_main.main(pyi_config, spec_file, **opts.__dict__)
    
      File "e:\python27\lib\site-packages\PyInstaller\building\build_main.py", line
    737, in main
        build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'
    ))
      File "e:\python27\lib\site-packages\PyInstaller\building\build_main.py", line
    680, in build
        exec(text, spec_namespace)
      File "<string>", line 16, in <module>
      File "e:\python27\lib\site-packages\PyInstaller\building\build_main.py", line
    207, in __init__
        self.__postinit__()
      File "e:\python27\lib\site-packages\PyInstaller\building\datastruct.py", line
    156, in __postinit__
        self.assemble()
      File "e:\python27\lib\site-packages\PyInstaller\building\build_main.py", line
    468, in assemble
        ctypes_binaries, ctypes_warnings = scan_code_for_ctypes(co)
      File "e:\python27\lib\site-packages\PyInstaller\depend\utils.py", line 164, in
     scan_code_for_ctypes
        nested_binaries, nested_warnings = scan_code_for_ctypes(c)
      File "e:\python27\lib\site-packages\PyInstaller\depend\utils.py", line 182, in
     scan_code_for_ctypes
        elif binary != os.path.basename(binary):
      File "e:\python27\lib\ntpath.py", line 208, in basename
        return split(p)[1]
      File "e:\python27\lib\ntpath.py", line 180, in split
        d, p = splitdrive(p)
      File "e:\python27\lib\ntpath.py", line 116, in splitdrive
        normp = p.replace(altsep, sep)
    AttributeError: 'tuple' object has no attribute 'replace'

**Comment by HeartUnchange** [comment]
i use the bare filenames but it still failed.so i think the problem is not caused by absolute paths.
this is an example without class:

    #!/usr/bin/env python
    #-*- coding:utf-8 -*-

    __author__ = 'Heartunchange'

    import ctypes

    if __name__ == '__main__':
        lib = ctypes.CDLL("library.dll")
        lib.love()

i use pyinstaller build it successfully.

**Comment by htgoebel** [comment]
This definitely is solved by  #1630, more precise by 989efa75c7cd07a50b53d5b1c53798ee0808b65c  Also see 11d779ff41ed71ec777fb5cc901d3145701d7e4e for a test-case and #1620.

@HeartUnchange Thanks for reporting. This was a very tricky one.
